### PR TITLE
create jwt

### PR DESCRIPTION
**Pull request by MrC1rcle**
create jwt with github token

### CODE REVIEW

File: internal/authentication/cert/id_rsa [hunk 1]
```diff
@@ -0,0 +1,51 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (The-TT-Hacker) [inline internal/authentication/cert/id_rsa]: generate this with terraform and store it in parameter store
https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key

File: internal/authentication/createjwt.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+func NewJWT(privateKey []byte, publicKey []byte) JWT {
+	return JWT{
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+}
+
+
+func CreateToken(ttl time.Duration, gittoken string) (string, error) {
+
```
Comment (The-TT-Hacker) [inline internal/authentication/createjwt.go]: dont have blank lines at the start of functions

File: internal/authentication/gettoken.go [hunk 3]
```diff
@@ -56,3 +63,47 @@ func GetToken(url string) (*Token, error) {
 	}
 	return &token, nil
 }
+
+func GetUserInfo(token *Token) (*GitHubUserInfo, error) {
+
```
Comment (The-TT-Hacker) [inline internal/authentication/gettoken.go]: remove blank line

File: internal/authentication/createjwt.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	publicKey  []byte
+}
+ 
+func NewJWT(privateKey []byte, publicKey []byte) JWT {
+	return JWT{
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+}
+
```
Comment (The-TT-Hacker) [inline internal/authentication/createjwt.go:23]: only need one line between the functions

File: internal/authentication/createjwt.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+func CreateToken(ttl time.Duration, gittoken string) (string, error) {
+
+	prvKey, err := ioutil.ReadFile("internal/authentication/cert/id_rsa")
+	if err != nil {
+		logger.Error(err.Error())
+			return "", err
+	}
+	pubKey, err := ioutil.ReadFile("internal/authentication/cert/id_rsa.pub")
```
Comment (The-TT-Hacker) [inline internal/authentication/createjwt.go]: fetch these from parameter store

File: internal/authentication/cert/id_rsa (hunk 1 above)
Comment (The-TT-Hacker) [reply internal/authentication/cert/id_rsa]: you'll want to reset this out, its never a good idea to have private keys in the repo

Comment (The-TT-Hacker) [reply internal/authentication/cert/id_rsa]: never do this again

Comment (MrC1rcle) [reply internal/authentication/cert/id_rsa]: yep it will not happen

File: internal/authentication/createjwt.go (hunk 5 above)
Comment (MrC1rcle) [reply internal/authentication/createjwt.go]: how to fetch from parameter? I wonder if we can store the client id and secret in the parameter too

File: internal/authentication/cert/id_rsa (hunk 1 above)
Comment (The-TT-Hacker) [reply internal/authentication/cert/id_rsa]: good man

File: internal/authentication/createjwt.go (hunk 5 above)
Comment (The-TT-Hacker) [reply internal/authentication/createjwt.go]: using the aws sdk
https://github.com/aws/aws-sdk-go-v2

File: frontend/public/settings.json [hunk 6]
```diff
@@ -1,4 +1,4 @@
 {
   "backend": "http://localhost:8080",
-  "client_id": "c2f25de6e7cfc7713944"
+  "client_id": "94143fe4a712d77c2983"
```
Comment (The-TT-Hacker) [inline frontend/public/settings.json]: maybe lets put a dummy value here instead of an actual value

File: internal/authentication/gettoken.go [hunk 7]
```diff
@@ -14,7 +21,7 @@ type Conf struct {
 }
 
 var conf = Conf{
-	ClientID:     "", // fill in with your id before test
+	ClientID:     "", // fill in with your id before test,
```
Comment (The-TT-Hacker) [inline internal/authentication/gettoken.go]: read these from environment variables which you can set in terraform

File: internal/authentication/gettoken.go [hunk 8]
```diff
@@ -14,7 +21,7 @@ type Conf struct {
 }
 
 var conf = Conf{
-	ClientID:     "", // fill in with your id before test
+	ClientID:     "", // fill in with your id before test, 
 	ClientSecret: "", // fill in with your secret before test
```
Comment (The-TT-Hacker) [inline internal/authentication/gettoken.go:26]: get the secret from parameter store

File: internal/authentication/gettoken.go [hunk 9]
```diff
@@ -14,7 +21,7 @@ type Conf struct {
...
 	RedirectURL:  "http://localhost:8080/login",
```
Comment (The-TT-Hacker) [inline internal/authentication/gettoken.go]: this should also be an environment variable you can set in terraform

File: internal/endpoints/auth.go [hunk 10]
```diff
@@ -25,14 +26,22 @@ func Authentication(service *web.Service) {
 
 		var err error
 		var code = in.Code
+		fmt.Println("code: ", code)
```
Comment (The-TT-Hacker) [inline internal/endpoints/auth.go]: dont print the code

File: internal/endpoints/auth.go [hunk 11]
```diff
@@ -25,14 +26,22 @@ func Authentication(service *web.Service) {
...
 		var tokenAuthURL = authentication.GetTokenAuthURL(code)
 		var token *authentication.Token
 		if token, err = authentication.GetToken(tokenAuthURL); err != nil {
 			logger.Error(err.Error())
 			return err
 		}
+		fmt.Println("token: ", token)
```
Comment (The-TT-Hacker) [inline internal/endpoints/auth.go]: dont print the token

File: internal/endpoints/auth.go (hunk 10 above)
Comment (MrC1rcle) [reply internal/endpoints/auth.go]: this is for testing, ill delete them before merge to main

File: terraform/main/rsakey.tf [hunk 12]
```diff
@@ -0,0 +1,11 @@
+resource "tls_private_key" "RSAprikey" {
```
Comment (The-TT-Hacker) [inline terraform/main/rsakey.tf]: use snake case for resource names

File: terraform/main/rsakey.tf [hunk 13]
```diff
@@ -0,0 +1,11 @@
+resource "tls_private_key" "rsa_prikey" {
+  algorithm = "RSA"
+  rsa_bits  = 4096
+}
+
+resource "aws_ssm_parameter" "private_key" {
+  name      = "prikey"
```
Comment (The-TT-Hacker) [inline terraform/main/rsakey.tf]: add the deployment id prefix here

File: internal/authentication/createjwt.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+}
+ 
+func NewJWT(privateKey []byte, publicKey []byte) JWT {
+	return JWT{
+		privateKey: privateKey,
+		publicKey:  publicKey,
+	}
+}
+
+func CreateToken(ttl time.Duration, gittoken string) (string, error) {
```
Comment (The-TT-Hacker) [inline internal/authentication/createjwt.go]: 1. this function is creating a private key not a token, they are different
2. you dont need to add the BEING PRIVATE KEY thing, the terraform private_key_pem value has it already

File: internal/authentication/getparameter.go [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    "context"
+    "dyno/internal/logger"
+	"github.com/aws/aws-sdk-go-v2/aws"
+    "github.com/aws/aws-sdk-go-v2/config"
+    "github.com/aws/aws-sdk-go-v2/service/ssm"
+)
+
+func getSSMParameterValue(parameterName string) (string){
+
+    cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("ap-southeast-2"))
```
Comment (The-TT-Hacker) [inline internal/authentication/getparameter.go]: you dont need to set the region, it should be done by default through the lambda environment variables

File: terraform/main/rsakey.tf (hunk 13 above)
Comment (MrC1rcle) [reply terraform/main/rsakey.tf]: what is a deployment id prefix? like "jwt_prikey"?

File: internal/authentication/getparameter.go (hunk 15 above)
Comment (MrC1rcle) [reply internal/authentication/getparameter.go]: so i cant just delete the region part?

File: terraform/main/rsakey.tf (hunk 13 above)
Comment (The-TT-Hacker) [reply terraform/main/rsakey.tf]: `"prikey"` -> `"${var.deployment_id}-prikey"`

File: internal/authentication/gettoken.go (hunk 8 above)
Comment (The-TT-Hacker) [reply internal/authentication/gettoken.go:26]: take in client id and secret in terraform variables. add them to terraform/variables.tf. Then store them in parameter store like how you've done it for the private key.

File: internal/authentication/gettoken.go (hunk 9 above)
Comment (The-TT-Hacker) [reply internal/authentication/gettoken.go]: theres a cf_endpoint output from the aws_only terraform you can use

Comment (The-TT-Hacker) [reply internal/authentication/gettoken.go]: or aws_cloudfront_distribution.frontend.domain_name

File: terraform/main/rsakey.tf (hunk 13 above)
Comment (MrC1rcle) [reply terraform/main/rsakey.tf]: it's easy. and do you mind explain why we need to do this?

File: internal/authentication/getparameter.go (hunk 15 above)
Comment (The-TT-Hacker) [reply internal/authentication/getparameter.go]: i think so

File: terraform/main/rsakey.tf (hunk 13 above)
Comment (The-TT-Hacker) [reply terraform/main/rsakey.tf]: so if multiple people are deploying the terraform there isnt a name clash

**Comment by The-TT-Hacker** [comment]
remove the empty id_rsa files

File: terraform/main/rsakey.tf (hunk 13 above)
Comment (MrC1rcle) [reply terraform/main/rsakey.tf]: ok i got it. thanks

File: frontend/package-lock.json [hunk 16]
```diff
@@ -1,16283 +1,8 @@
 {
   "name": "golambda-ui",
   "version": "0.1.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (The-TT-Hacker) [inline frontend/package-lock.json]: you are on an old version of node/npm. you shouldnt be changing this file

Comment (The-TT-Hacker) [reply frontend/package-lock.json]: update to node 16.16 and this wont happen. i would recommend install node version manager https://github.com/nvm-sh/nvm

**Review by The-TT-Hacker — APPROVED** [review]

**Pull request merged**